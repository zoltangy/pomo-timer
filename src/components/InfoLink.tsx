import React, { useState, useEffect } from "react";
import {
  IconButton,
  Link,
  Dialog,
  Typography,
  DialogContent,
  DialogTitle,
  DialogActions,
  Button,
  Grid,
} from "@material-ui/core";
import CloseIcon from "@material-ui/icons/Close";
import InfoOutlinedIcon from "@material-ui/icons/InfoOutlined";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles((theme) => ({
  icon: {
    fontSize: 30,
  },
}));

function OfflineDialog() {
  const classes = useStyles();
  const [open, setOpen] = useState(false);
  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };
  return (
    <>
      <div>
        <IconButton
          edge="start"
          aria-label="Info link offline"
          color="primary"
          size="small"
          onClick={handleOpen}
        >
          <InfoOutlinedIcon className={classes.icon} />
        </IconButton>
      </div>
      <Dialog
        open={open}
        onClose={handleClose}
        aria-label="Offline info dialog"
        aria-describedby="offline-link-description"
      >
        <DialogTitle>
          <Grid container spacing={1} justify="space-between">
            <Typography variant="h6" component="span">
              Oops!
            </Typography>
            <IconButton
              edge="end"
              aria-label="close dialog"
              color="primary"
              size="small"
              onClick={handleClose}
            >
              <CloseIcon className={classes.icon} />
            </IconButton>
          </Grid>
        </DialogTitle>
        <DialogContent dividers>
          <Typography id="offline-link-description" align="justify" gutterBottom>
            Your device seems to be offline, so the information page can't be accessed at the moment, but the
            rest of the app is fully functional even when offline.
          </Typography>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleClose}>Ok</Button>
        </DialogActions>
      </Dialog>
    </>
  );
}

function OnlineInfo() {
  const classes = useStyles();
  return (
    <Link href="https://en.wikipedia.org/wiki/Pomodoro_Technique" target="_blank" rel="noopener">
      <IconButton edge="start" aria-label="Info link online" color="primary" size="small">
        <InfoOutlinedIcon className={classes.icon} />
      </IconButton>
    </Link>
  );
}

const InfoLink: React.FC = () => {
  const [online, setOnline] = useState<boolean>(true);

  useEffect(() => {
    function online() {
      setOnline(true);
    }
    function offline() {
      setOnline(false);
    }

    window.addEventListener("offline", offline);
    window.addEventListener("online", online);
    return function cleanup() {
      window.removeEventListener("offline", offline);
      window.removeEventListener("online", online);
    };
  }, []);

  return <>{online ? <OnlineInfo /> : <OfflineDialog />}</>;
};

export default InfoLink;
