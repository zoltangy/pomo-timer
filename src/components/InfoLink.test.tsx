import React from "react";
import { render, act, fireEvent } from "@testing-library/react";
import InfoLink from "./InfoLink";

describe("InfoLink tests", () => {
  it("shows component when online/offline", async () => {
    const { asFragment, queryByLabelText, getByLabelText, baseElement } = render(<InfoLink />);

    expect(queryByLabelText("Info link online")).toBeTruthy();
    expect(queryByLabelText("Info link offline")).toBeNull();
    expect(asFragment()).toMatchSnapshot();

    act(() => {
      window.dispatchEvent(new Event("offline"));
    });

    expect(queryByLabelText("Info link online")).toBeNull();
    expect(queryByLabelText("Info link offline")).toBeTruthy();
    expect(asFragment()).toMatchSnapshot();

    fireEvent.click(getByLabelText("Info link offline"));
    expect(queryByLabelText("Offline info dialog")).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
  });
});
