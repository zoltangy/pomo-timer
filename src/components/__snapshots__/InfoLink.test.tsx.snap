// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoLink tests shows component when online/offline 1`] = `
<DocumentFragment>
  <a
    class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
    href="https://en.wikipedia.org/wiki/Pomodoro_Technique"
    rel="noopener"
    target="_blank"
  >
    <button
      aria-label="Info link online"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-icon-1"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </a>
</DocumentFragment>
`;

exports[`InfoLink tests shows component when online/offline 2`] = `
<DocumentFragment>
  <div>
    <button
      aria-label="Info link offline"
      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall MuiIconButton-edgeStart"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root makeStyles-icon-1"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</DocumentFragment>
`;

exports[`InfoLink tests shows component when online/offline 3`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  >
    <div>
      <button
        aria-label="Info link offline"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall MuiIconButton-edgeStart"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-icon-1"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    aria-label="Offline info dialog"
    class="MuiDialog-root"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper"
      role="none presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-describedby="offline-link-description"
        class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
        role="dialog"
      >
        <div
          class="MuiDialogTitle-root"
        >
          <h2
            class="MuiTypography-root MuiTypography-h6"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-xs-space-between"
            >
              <span
                class="MuiTypography-root MuiTypography-h6"
              >
                Oops!
              </span>
              <button
                aria-label="close dialog"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeSmall MuiIconButton-edgeEnd"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root makeStyles-icon-1"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </h2>
        </div>
        <div
          class="MuiDialogContent-root MuiDialogContent-dividers"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiTypography-alignJustify"
            id="offline-link-description"
          >
            Your device seems to be offline, so the information page can't be accessed at the moment, but the rest of the app is fully functional even when offline.
          </p>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Ok
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
